h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1735,7 +1744,9 @@ public int compare(Chunk o1, Chunk o2) {
                 }
             }
             if (!queue.isEmpty()) {
-                result = queue;
+                ArrayList<Chunk> list = new ArrayList<>(queue);
```

Comment:
This collection has sort, remove and iterate operations, should we use ArrayList or LinkedList?

Reply:
It looks like `ArrayList` is significantly cheaper here.

Reply:
@katzyn @andreitokar According my test, here using ArrayList is overhead and bad.
# Test result
```shell
$java ListSIRPerfTest array 200000
10:04:32.163 Add start
10:04:32.169 Sort start
10:04:32.267 Iterate & remove start
10:04:36.076 time: 3913ms, items: 200000, type: java.util.ArrayList

$java ListSIRPerfTest linked 200000
10:04:38.474 Add start
10:04:38.486 Sort start
10:04:38.590 Iterate & remove start
10:04:38.600 time: 126ms, items: 200000, type: java.util.LinkedList
```
**LinkedList is 10 times faster than ArrayList.**

# My test case
```java
import java.util.*;
import java.text.*;

// List perf test for Sort, Iterate, and Remove.
public class ListSIRPerfTest {

  public static void main(String args[]){
    String type = "array";
	int n = 1000000;
	DateFormat df = new SimpleDateFormat("HH:mm:ss.SSS");

	if (args.length > 0) {
	  type = args[0];
	}
	if (args.length > 1) {
	  n = Integer.parseInt(args[1]);
	}
	PriorityQueue<Integer> queue = new PriorityQueue<>();
	Random rand = new Random();
	for(int i = 0; i < n; ++i){
		queue.offer(rand.nextInt(n+1));
	}

    long ts = System.currentTimeMillis();
	// 1. Add
	System.out.println(String.format("%s Add start", df.format(new Date())));
	List<Integer> list;
	if("linked".equals(type)){
	  list = new LinkedList<>(queue);
	}else{
	  list = new ArrayList<>(queue);
	}
	
        // 2. Sort
	System.out.println(String.format("%s Sort start", df.format(new Date())));
	Collections.sort(list);

	// 3. Iterate & remove
	System.out.println(String.format("%s Iterate & remove start", df.format(new Date())));
	for(Iterator<Integer> i = list.iterator(); i.hasNext();){
	  Integer val = i.next();
	  i.remove();
	}

	System.out.println(String.format("%s time: %dms, items: %d, type: %s", 
	  df.format(new Date()), (System.currentTimeMillis() - ts), n, list.getClass().getName()));
  }

}
```

Reply:
How your test case is related with usage of the collection in this specific place? Typically there is only one deletion.

And you really should use JMH for such tests.

Reply:
@katzyn the ArrayList is used to move chunk in mvstore and  element is removed if moving successfully in iterator.

Reply:
Only when we have many of them, but we don't. Other operations are faster in `ArrayList`.

Do you have a test case with many `remove()` operations in a large list? I can't reproduce this situation.

Reply:
@katzyn I only discuss the mvstore moving chunks. It's easy to calculate the quantity of chunks moved.

Reply:
Number of removals may be zero or may be all of them, but most of the time is expected to be small. It is a no-brainer, that big ArrayList is performing poorly on removals, but question is how big and how many removals, and @little-pan, your test may or may not be realistic.
Regardless of that I think difference between ArrayList and LinkedList usage here is going to be just a drop in the bucket, compare to the cost of actual movement of each Chunk. Number of such movements should be main goal of optimization, IMHO.
Just last night I realized that after introduction of reserved area, I can get away without any removals at all, because those two loops can be combined. That would make thisdiscussion somewhat irrelevant.

Reply:
> Just last night I realized that after introduction of reserved area, I can get away without any removals at all, because those two loops can be combined.

@andreitokar If two loops can be combined, it's very good.

Reply:
done